nu_stats/fqutils.py

```diff
@@ -0,0 +1,772 @@
+            low_E = np.log10(E) < midE
+            otherEi = self._other_index(E_index, low_E)
+
+            if otherEi is None:
+                f1 = f0
+            else:
+                f1 = self._likelihood[i_index, otherEi]
+
+            dE = ((np.log10(E) - midE) / self._Ebin_width)
+            dE -= 2*dE*low_E # switch sign if to the low
+
+            if interpol == 1:
+                # linear interpolation
+                lik = ((1-dE)*f0 + (dE)*f1)/ self._Nbins
+                return max(lik, 1e-10)
+
+            elif interpol == 2:
+                # bilinear interpolation
+                # f00, f01, f10, f11
+                f = [f0, f0, f0, f0]
+                midi = self._index_bins[i_index] + self._ibin_width/2
+                low_i = new_index < midi
+                otherii = self._other_index(i_index, low_i)
+
+                if otherEi is not None:
+                    f[1] = self._likelihood[i_index, otherEi]
+                    f[3] = f[1] # will be kept if otherii is None
+
+                if otherii is not None:
+                    f[2] = self._likelihood[otherii, E_index]
+                    f[3] = f[2] # will be kept if otherEi is None
+
+                if (otherEi is not None) and (otherii is not None):
+                    f[3] = self._likelihood[otherii, otherEi]
+
+                di = ((new_index - midi) / self._ibin_width)
+                di -= 2*di*low_i
+
+                lik = (f[0] * (1-dE) * (1-di)
+                    + f[1] * dE * (1-di)
+                    + f[2] * (1-dE) * di
+                    + f[3] * dE * di) / self._Nbins
+
+            return max(lik, 1e-10)
+
+        else:
+            return self._likelihood[i_index, E_index]
+
+    def plot_pdf_at_idx(self, index, interp=2, scaled_plot = True, **kwargs):
+        Es = np.logspace(np.log10(self._min_E),
+                 np.log10(self._max_E), 10000)
+        ll = [self(E, index, interp) for E in Es]
+        if interp and not scaled_plot:
+            ll = [l*self._Nbins for l in ll]
+        p = plt.plot(Es, ll, **kwargs)
+        plt.xscale('log')
+        return p
+
+    def plot_pdf_at_E(self, E, interp=2, scaled_plot = True, **kwargs):
+        idx = np.linspace(self._min_index,
+                 self._max_index,10000)
+        ll = [self(E, index, interp) for index in idx]
+        if interp and not scaled_plot:
+            ll = [l*self._Nbins for l in ll]
+        p = plt.plot(idx, ll, **kwargs)
+        return p
+
+    def plot_pdf_meshgrid(self, interp=2, **kwargs):
+        idx = np.linspace(self._min_index,
+                 self._max_index,100)
+        Es = np.logspace(np.log10(self._min_E),
+                 np.log10(self._max_E),100)
+        Es, idx = np.meshgrid(Es, idx)
+        vfunc = np.vectorize(self.__call__)
+        Z = vfunc(Es, idx, interp)
+        # np.savetxt("pdfmesh.csv", Z, delimiter=",")
+        plt.pcolormesh(Es, idx, Z, shading='auto', **kwargs)
+        plt.xscale('log')
```

Comment:
I feel like AtmosphericEnergyLikelihood and MarginalisedEnergyLikelihood could be the same class, or at least share a base class to avoid repeating lots of code.